### ISSUE TITLE
Unable to create function app against existing consumption plan

### ISSUE THREAD

**Issue by Jameskmonger** [Azure/azure-cli#8388]
**Describe the bug**
It is not possible to create a function app against a specified consumption app service plan. The following error occurs:

> There was a conflict. AlwaysOn cannot be set for this site as the plan does not allow it.

I think that setting `--consumption-plan-location` disables `AlwaysOn`, but `--plan` doesn't.

**To Reproduce**
1. Create a consumption plan (see "Additional context" section)
2. Try to create a function app against the plan created in step 1:
    `az functionapp create --resource-group myresourcegroup --storage-account mystorageaccount --name myfunction --plan PlanFromStepOne`
3. Get the error:
    >There was a conflict. AlwaysOn cannot be set for this site as the plan does not allow it.

**Expected behavior**
It should create the function app against the specified plan, and not set "AlwaysOn" if the plan does not support it.

**Environment summary**
Install method: MSI
CLI version: 2.0.52
OS version: Windows 10, Version 1803, OS Build 17134.523
Shell type: cmd

**Additional context**
You can create a consumption app service plan in two ways.

The first way will create a plan called "NorthEuropePlan" with no option to specify the name (it's also part of creating a function app, rather than a discrete step):

```
az functionapp create
    --resource-group myresourcegroup 
    --storage-account mystorageaccount 
    --name myfunctionapp
    --runtime node
    --consumption-plan-location northeurope
```

The second is to use the `az resource` command. I consider this a **much** better approach as it is a discrete step, and also lets you specify the name:

```
az resource create
    --resource-group myresourcegroup
    --name myconsumptionplan
    --resource-type Microsoft.web/serverfarms
    --is-full-object
    --properties "{\"location\":\"northeurope\",\"sku\":{\"name\":\"Y1\",\"tier\":\"Dynamic\"}}"
```

**Comment by pfranco22** [comment]
Hi, if you are using an Consumption Plan (Dynamic) you only need to set --consumption-plan-location, not --plan. 
Consumption plans does not allow Always-On. If you create an App Service plan, then you can use Always On.

**Comment by Jameskmonger** [comment]
@pfranco22 What if I already have a consumption plan and I want to create a new app against that plan? `--consumption-plan-location` does not meet that requirement

**Comment by uberchris2** [comment]
@Jameskmonger it seems like if you specify --consumption-plan-location, the function app will use the consumption plan in that region. 

However, the [documentation](https://docs.microsoft.com/en-us/azure/azure-functions/scripts/functions-cli-create-app-service-plan#sample-script) and error message does not imply this, and both should be updated to clarify.

**Comment by bigdatamoore** [comment]
I just ran into this.   @uberchris2 is right, if you don't specify a plan, but instead give the consumption-plan-location of your current plan, it creates the functionapp in the existing plan.  not what I expected...

**Comment by Vannevelj** [comment]
Had the same issue but using `--consumption-plan-location` didn't work and I'd just end up with `WestEuropePlan` and `WestEuropePlan2`. 

In order to get around this, I've changed the tier from `F1` to `B1` (the free tier does not allow the `AlwaysOn` setting).

**Comment by ManuelGalindo** [comment]
I have the same issue with an App Service, when I try to create a Service Plan and then assign to a shared SKU App Service I got this message

`az` appservice plan create --name "MyAwesomePlan" --resource-group "Super" --location "canadacentral" --sku SHARED
`az` webapp create -g "Super" -p "MyAwesomePlan" -n "MySuperWebApp" --verbose

Is possible to still doing so or cancel the property of Always On from Az CLI?

**Comment by TheBlueSky** [comment]
I have the same issue and I'm using the latest version of the CLI (as of this writing it's v2.0.69).

Is there a way to create a `functionapp` without setting "AlwaysOn"?

**Comment by David-Stark** [comment]
> Hi, if you are using an Consumption Plan (Dynamic) you only need to set --consumption-plan-location, not --plan.
> Consumption plans does not allow Always-On. If you create an App Service plan, then you can use Always On.

Sorry, but this method creates a very bad name for the consumption plan. If you have 5 azure functions in north europe, they will all be named NorthEuropePlan.

**Comment by mrscommentor** [comment]
This is extremely bad, I should be able to use an F1 free tier plan, which I can through the portal, but via the az cli it assumes for some reason that always on is true even though I don't need always on. Also using consumption based ruins our naming standards as we can't control the name of the auto generated plan, so that is not an option either. Please default to always on false, or give us the option to set that flag via the cli (az functionapp create).

**Comment by IanKemp** [comment]
Come ON, Microsoft. How difficult is it to fix this?

**Comment by hentie** [comment]
I've also ran into this issue.  Creating a functionapp with Consumption Plan with a name that conforms to our naming convention is a breeze in ARM, but to create it with Azure CLI, I had to do a horrible workaround:
1. Create a consumption plan like @Jameskmonger described

`az resource create
    --resource-group myresourcegroup
    --name myconsumptionplan
    --resource-type Microsoft.web/serverfarms
    --is-full-object
    --properties "{\"location\":\"westeurope\",\"sku\":{\"name\":\"Y1\",\"tier\":\"Dynamic\"}}"`

2. Create a function app with the --consumption-plan-location i.e.

`az functionapp create 
        --resource-group myresourcegroup
        --name myfunctionapp
        --consumption-plan-location westeurope`

3. Update the functionapp to use the plan created in step 1  i.e.

`az functionapp update
        --resource-group myresourcegroup
        --name myfunctionapp
        --plan myconsumptionplan`

4. Delete the consumption plan auto generated in step 2 i.e. (I prefer az resource delete for this)

`az resource delete 
--name WestEurope 
--resource-group myresourcegroup` 


I've also found that ver 2.0.8 Azure CLI does not give the "Always On" issue that @Jameskmonger described, but 2.3.1 and 2.4.0 do.  But regardless, it just makes sense to be able to specify a newly created consumption plan name for a functionapp in the Azure-CLI and in my opinion should be implemented a.s.a.p.

**Comment by vikaskk13** [comment]
I also ran into this issue. and followed the steps mentioned by @hentie .. But what I noticed is that,
after step 2 If I look at the function app, it has automatically linked to the existing consumption app service plan..

**Comment by TulyOpt** [comment]
Hi, any update ?

**Comment by lilyjma** [comment]
@ahmedelnably, @fabiocav please take a look at the issue.

**Comment by hentie** [comment]
> I also ran into this issue. and followed the steps mentioned by @hentie .. But what I noticed is that,
> after step 2 If I look at the function app, it has automatically linked to the existing consumption app service plan..

@vikaskk13 That is exactly why I call it a horrible work-around.  You have to temporarily create a resource just to delete it again :(

**Comment by graphicsxp** [comment]
Having the same issue. Why is this still not assigned for fixing?

**Comment by yonzhan** [comment]
functions service team should look into this.

**Comment by Silvenga** [comment]
Wonder if you tried your workaround recently @hentie - it looks like it's causing the Azure management API to throw 500's when trying to create a new function app. Or maybe Azure is just having a bad day - can never really tell. 🤷

**Comment by hentie** [comment]
> Wonder if you tried your workaround recently @hentie - it looks like it's causing the Azure management API to throw 500's when trying to create a new function app. Or maybe Azure is just having a bad day - can never really tell. 🤷

Hi @Silvenga , I just tested it this morning and the '_terrible workaround_' is still working on my side.  I was using Azure CLI version 2.4.0 and deploying the the resources to West Europe region for the test.

**Comment by Silvenga** [comment]
Perfect, thanks for checking @hentie!

**Comment by Silvenga** [comment]
Spoke with Azure Support - trying to create a Linux consumption plan with a custom name will return a 500 error, which seems to be a "works as designed". I may be in the minority, but that seems a little, uhm, wrong.

So it looks like something is fundamentally broken in Azure that needs to be fixed before this feature can be supported.

**Comment by IOrlandoni** [comment]
Still broken...

**Comment by igor-mamaliga** [comment]
I had the same issue and have achieved the result by manually creating the function from the Azure portal.
Since the portal allows the creation of a function for F1:Free app service plan I am wondering how do they do it ?

**Comment by furey** [comment]
Something like a `--consumption-plan-name` option would be a massive help.

🤓 
🙏

**Comment by norvegec** [comment]
2 years gone by and this is still open?

**Comment by fabiocav** [comment]
@norvegec I hope the fact that a lot of the work done by the team is in the open provides good evidence that the team hasn't been addressing problems. Issues do need to be prioritized and, unfortunately, as much as we would love to, particularly with the challenges we all had to deal with this previous year, we can't always address them in a timely manner. I do agree that there should have been a response and proper triage on this, and I'm sorry this didn't happen (it looks like this was created prior to some processes that would trigger the triage and somehow, feel through). I'm flagging this for discussion in the next triage meeting.

@apawast can we please add this to our next triage list? We may need to loop in Web Apps (Antares) folks as the plan for consumption differs from a traditional plan and there might be some limitations there. At the very least, we need to have clear documentation and guidance on this so we can have closure on this issue.

**Comment by norvegec** [comment]
@fabiocav agree, given 16 👍 this looks like low-priority issue. My point is to either provide an estimate for a fix, mark @hentie solution as recommended or an alternative solution(s) and close this.

**Comment by IanKemp** [comment]
@fabiocav Problems have been addressed, yes, but they honestly haven't been addressed as quickly as they should have been. I don't blame you or other teams that are similarly slow to provide feedback (NuGet comes to mind), because I'm quite certain that the reason is that you simply don't have enough people to deal with the number of issues reported. That's a management problem not a dev team problem, but unfortunately since the dev team (you) are the only points of contact for customers, you are the ones feeling the pain.

And that sucks, but... as a paying customer none of the above should be my responsibility or concern. If Microsoft is going to understaff its product teams in order to maximize shareholder profits that's their choice, but then they shouldn't be surprised when customers get unhappy and/or start jumping ship to AWS.

**Comment by navba-MSFT** [comment]
@Jameskmonger Apologies for the late reply. I am able to create function app with existing consumption plan now. Did you get a chance to test this and check? This issue is open for quite sometime now. Could you please let us know if you need any assistance on this ? Awaiting your reply.

**Comment by hentie** [comment]
> @Jameskmonger Apologies for the late reply. I am able to create function app with existing consumption plan now. Did you get a chance to test this and check? This issue is open for quite sometime now. Could you please let us know if you need any assistance on this ? Awaiting your reply.

> Hi, we're sending this friendly reminder because we haven't heard back from you in a while. We need more information about this issue to help address it. Please be sure to give us your input within the next **7 days**. If we don't hear back from you within **14 days** of this comment the issue will be automatically closed. Thank you!

The issue still exists:  You cannot specify the name of the consumption plan when creating a new function app.  The name of the consumption plan is based on the location, which is not desired.

**Comment by hentie** [comment]
The issue still exists: You cannot specify the name of the consumption plan
when creating a new function app. The name of the consumption plan is based
on the location, which is not desired.

**Comment by davesouthard** [comment]
It would be nice to have this fixed.

**Comment by navba-MSFT** [comment]
@Francisco-Gamino Could you please look into this and provide an update once you get a chance ? Awaiting your reply.

**Comment by Francisco-Gamino** [comment]
Hello @navba-MSFT -- I do not work on the `Az CLI`. Adding @panchagnula @gzuber who might be able to help.

**Comment by constantine1998** [comment]
The issue still exists, The name of the consumption cant be specified while creating the function, and the name of the consumption app is still based on location. Any Update on this!?

**Comment by panchagnula** [comment]
> The issue still exists, The name of the consumption cant be specified while creating the function, and the name of the consumption app is still based on location. Any Update on this!?

This is not a CLI issue  - I don't think this is possible on any client during functionapp create. @fabiocav can you comment on this?

**Comment by thomasddn** [comment]
> This is not a CLI issue - I don't think this is possible on any client during functionapp create. @fabiocav can you comment on this?

It is possible with an ARM template.

**Comment by Azrrael** [comment]
We never had this issue before, and now it happens but intermittently. When creating a function using:

`az functionapp create --consumption-plan-location $1 --name $2 --os-type $3 --resource-group $4 --runtime $5 --storage-account $6 --functions-version $7 --app-insights $8`

sometimes it will create the function in a preexisting Function App on $4 resource group, sometimes it will create a new Function App called "EastUS2Plan" -- looks like it depends on the region. This is really bad as it means that random Function Apps might appear for no reason.

**Comment by constantine1998** [comment]
Strange to see a new function app getting created! It creates a consumption plan("EastUS2Plan") based on your location name, but never creates a new function app("EastUS2Plan") !!

**Comment by Jameskmonger** [comment]
@navba-MSFT I don't use Azure anymore, I have changed jobs in the time since posting this thread, so I'm not able to give any more info. Hopefully the other users here can provide some

**Comment by paul-sk** [comment]
Create consumption plan using 'az resource create --name myconsumptionplan ...' as described above in the resource group where function app will be created. Then create function app with '...--consumption-plan-location mylocation...' in the resource group. This function app will use myconsumptionplan created, but not <mylocation>Plan.  Just tested.

**Comment by uglywarthog** [comment]
> Create consumption plan using 'az resource create --name myconsumptionplan ...' as described above in the resource group where function app will be created. Then create function app with '...--consumption-plan-location mylocation...' in the resource group. This function app will use myconsumptionplan created, but not mylocationPlan. Just tested.

Yes, this is how it works. I have been struggling with creating and using proper consumption plan for functions and just have found out that if you create a plan in given location, the if you create functionapp in the same location it will use the plan.

Not how it should work though 😕

**Comment by MichielDeKoninck** [comment]
This issue still exists: I cannot create a Linux based FunctionApp that is linked to an app service consumption plan (SKU: Y1). I currently see no way of elegantly fixing this. 

Why are we unable to indicate that the property `AlwaysOn` should not be set for a FunctionApp we want to create on an existing consumption plan?

![image](https://user-images.githubusercontent.com/26138002/206142762-90e4a1fe-eae2-498f-875a-2624ff459e28.png)


![image](https://user-images.githubusercontent.com/26138002/206142612-81267f4a-efd8-46e3-8947-e035f558ba96.png)

**Comment by navba-MSFT** [comment]
@Stefanus Hinardi, @Francisco-Gamino Could you please look into this once you get a chance ? Awaiting your reply.

**Comment by AndreiZhukouski** [comment]
What a shame, Microsoft! Still not resolved! I have a bunch of Azure Function apps to deploy (20+), each of them has its own App Service Plan, Storage Account, and so on. Also we have naming convention defined (!!), so the name CentralUSPlan is not suitable for us.
Make it possible either turn of the AlwaysOn feature during deployment or let users select the name of the plan to create!

**Comment by Francisco-Gamino** [comment]
Adding Functions Az CLI team @amamounelsayed, @kaibocai, @shreyas-gopalakrishna, @kamperiadis, @apawast.

**Comment by CornerstoneII** [comment]
Issue:

While using Terraform to create Azure Function Apps alongside an App Service Plan with SKU EP1, I consistently encountered errors indicating conflicts with setting the Always On feature. However, the error messages provided were not sufficiently informative until I activated Terraform debug logging with TF_LOG=DEBUG.

After some investigation, I discovered that turning off Always On allowed the Function Apps to be created successfully, indicating a potential issue with the provider's handling of this setting for Function Apps using an EP1 SKU App Service Plan.

Expected behavior:

Given that the EP1 SKU supports the Always On feature, terraform should be able to successfully create Azure Function Apps with this setting enabled without encountering errors.

Steps to reproduce:

Create an App Service Plan with SKU EP1 using the azurerm_service_plan resource.
Attempt to create a Function App with Always On enabled using the azurerm_windows_function_app or azurerm_linux_function_app resource.
Environment:

Terraform version: [your terraform version]
AzureRM Provider version: [your AzureRM Provider version]
Additional context:

Here is a snippet of the configuration used:

```
resource "azurerm_linux_function_app" "linux_func_app" {
  for_each = local.linux_function_app_name

  name                       = ""
  location                   = var.location
  resource_group_name        = var.rg
  service_plan_id            = azurerm_service_plan.funapp_asp.id
  storage_account_name       = azurerm_storage_account.funapp_st_acc.name
  storage_account_access_key = azurerm_storage_account.funapp_st_acc.primary_access_key

  site_config {
    always_on = false
    application_stack {
      # node_version = "~16" # 12, 14, 16 and 18
      python_version = "3.9" # 3.11, 3.10, 3.9, 3.8 and 3.7
      # java_version = "17" # 8, 11 & 17
      # dotnet_version = "7.0" # 3.1, 6.0 and 7.0
    }
  }

  app_settings = {
    "APPINSIGHTS_INSTRUMENTATIONKEY" = azurerm_application_insights.appinsights.instrumentation_key
    "FUNCTIONS_WORKER_RUNTIME"       = "python"
    "FUNCTIONS_EXTENSION_VERSION"    = "~4"
  }
  depends_on = [azurerm_storage_account.funapp_st_acc]
}
```

As a workaround, setting Always On to false allows the Function Apps to be created successfully.

Could you please look into this issue and provide a fix or further guidance?

Thank you.

**Comment by ricardofalc** [comment]
> Issue:
> 
> While using Terraform to create Azure Function Apps alongside an App Service Plan with SKU EP1, I consistently encountered errors indicating conflicts with setting the Always On feature. However, the error messages provided were not sufficiently informative until I activated Terraform debug logging with TF_LOG=DEBUG.
> 
> After some investigation, I discovered that turning off Always On allowed the Function Apps to be created successfully, indicating a potential issue with the provider's handling of this setting for Function Apps using an EP1 SKU App Service Plan.
> 
> Expected behavior:
> 
> Given that the EP1 SKU supports the Always On feature, terraform should be able to successfully create Azure Function Apps with this setting enabled without encountering errors.
> 
> Steps to reproduce:
> 
> Create an App Service Plan with SKU EP1 using the azurerm_service_plan resource. Attempt to create a Function App with Always On enabled using the azurerm_windows_function_app or azurerm_linux_function_app resource. Environment:
> 
> Terraform version: [your terraform version] AzureRM Provider version: [your AzureRM Provider version] Additional context:
> 
> Here is a snippet of the configuration used:
> 
> ```
> resource "azurerm_linux_function_app" "linux_func_app" {
>   for_each = local.linux_function_app_name
> 
>   name                       = ""
>   location                   = var.location
>   resource_group_name        = var.rg
>   service_plan_id            = azurerm_service_plan.funapp_asp.id
>   storage_account_name       = azurerm_storage_account.funapp_st_acc.name
>   storage_account_access_key = azurerm_storage_account.funapp_st_acc.primary_access_key
> 
>   site_config {
>     always_on = false
>     application_stack {
>       # node_version = "~16" # 12, 14, 16 and 18
>       python_version = "3.9" # 3.11, 3.10, 3.9, 3.8 and 3.7
>       # java_version = "17" # 8, 11 & 17
>       # dotnet_version = "7.0" # 3.1, 6.0 and 7.0
>     }
>   }
> 
>   app_settings = {
>     "APPINSIGHTS_INSTRUMENTATIONKEY" = azurerm_application_insights.appinsights.instrumentation_key
>     "FUNCTIONS_WORKER_RUNTIME"       = "python"
>     "FUNCTIONS_EXTENSION_VERSION"    = "~4"
>   }
>   depends_on = [azurerm_storage_account.funapp_st_acc]
> }
> ```
> 
> As a workaround, setting Always On to false allows the Function Apps to be created successfully.
> 
> Could you please look into this issue and provide a fix or further guidance?
> 
> Thank you.

For me removing the entire  always_on option works, setting it to false OR true does not work anymore.

I get errors "AlwaysOn cannot be set for this site as the plan does not allow it." while running a consumption based plan which supports AlwaysOn according to the docs :)

**Comment by sicil1ano** [comment]
> I've also ran into this issue. Creating a functionapp with Consumption Plan with a name that conforms to our naming convention is a breeze in ARM, but to create it with Azure CLI, I had to do a horrible workaround:
> 
> 1. Create a consumption plan like @Jameskmonger described
> 
> `az resource create --resource-group myresourcegroup --name myconsumptionplan --resource-type Microsoft.web/serverfarms --is-full-object --properties "{\"location\":\"westeurope\",\"sku\":{\"name\":\"Y1\",\"tier\":\"Dynamic\"}}"`
> 
> 2. Create a function app with the --consumption-plan-location i.e.
> 
> `az functionapp create --resource-group myresourcegroup --name myfunctionapp --consumption-plan-location westeurope`
> 
> 3. Update the functionapp to use the plan created in step 1  i.e.
> 
> `az functionapp update --resource-group myresourcegroup --name myfunctionapp --plan myconsumptionplan`
> 
> 4. Delete the consumption plan auto generated in step 2 i.e. (I prefer az resource delete for this)
> 
> `az resource delete --name WestEurope --resource-group myresourcegroup`
> 
> I've also found that ver 2.0.8 Azure CLI does not give the "Always On" issue that @Jameskmonger described, but 2.3.1 and 2.4.0 do. But regardless, it just makes sense to be able to specify a newly created consumption plan name for a functionapp in the Azure-CLI and in my opinion should be implemented a.s.a.p.

Thanks for sharing this workaround. Any update on the issue?
